In [1]:
import warnings
warnings.filterwarnings("ignore")

import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import env
import os

In [2]:
def get_hvac_data():
    return pd.read_csv('hvac.csv')
    return df

In [3]:
df=get_hvac_data()

In [4]:
df.head()

Year        Region Type Units  AFUE  SEER  Ducted  Refrigerant  \
0  2021  Northeastern   AC    33   NaN  13.0     0.0        410.0   
1  2021  Northeastern   AC   326   NaN  13.0     0.0        410.0   
2  2021  Northeastern   AC   289   NaN  13.0     0.0        410.0   
3  2021  Northeastern   AC    31   NaN  13.0     0.0        410.0   
4  2021  Northeastern   AC    49   NaN  13.0     0.0        410.0   

  Cooling (Btu/hr) Heating (Btu/hr) MotorType HSPF FinalEstimate  EER FuelType  
0           18,000              NaN       NaN  NaN           165  NaN      NaN  
1           24,000              NaN       NaN  NaN         1,631  NaN      NaN  
2           30,000              NaN       NaN  NaN         1,446  NaN      NaN  
3           42,000              NaN       NaN  NaN           155  NaN      NaN  
4           48,000              NaN       NaN  NaN           245  NaN      NaN

# Summarize the data

In [6]:
df.head()

Year        Region Type Units  AFUE  SEER  Ducted  Refrigerant  \
0  2021  Northeastern   AC    33   NaN  13.0     0.0        410.0   
1  2021  Northeastern   AC   326   NaN  13.0     0.0        410.0   
2  2021  Northeastern   AC   289   NaN  13.0     0.0        410.0   
3  2021  Northeastern   AC    31   NaN  13.0     0.0        410.0   
4  2021  Northeastern   AC    49   NaN  13.0     0.0        410.0   

  Cooling (Btu/hr) Heating (Btu/hr) MotorType HSPF FinalEstimate  EER FuelType  
0           18,000              NaN       NaN  NaN           165  NaN      NaN  
1           24,000              NaN       NaN  NaN         1,631  NaN      NaN  
2           30,000              NaN       NaN  NaN         1,446  NaN      NaN  
3           42,000              NaN       NaN  NaN           155  NaN      NaN  
4           48,000              NaN       NaN  NaN           245  NaN      NaN

In [7]:
df.shape

(7815, 15)

In [5]:
df.describe()

Year         AFUE         SEER       Ducted  Refrigerant  \
count  7815.000000  2840.000000  4776.000000  4857.000000  4845.000000   
mean   2018.852335    88.250704    17.561139     0.437719   407.437358   
std       1.264162     5.980640     3.303248     0.496157    31.431542   
min    2017.000000    80.000000    13.000000     0.000000    22.000000   
25%    2018.000000    83.000000    15.000000     0.000000   410.000000   
50%    2019.000000    86.000000    17.000000     0.000000   410.000000   
75%    2020.000000    95.000000    20.000000     1.000000   410.000000   
max    2021.000000    99.000000    26.000000     1.000000   410.000000   

               EER  
count  3274.000000  
mean     11.808186  
std       1.636501  
min       9.000000  
25%      11.000000  
50%      12.000000  
75%      13.000000  
max      20.000000

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7815 entries, 0 to 7814
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Year              7815 non-null   int64  
 1   Region            7815 non-null   object 
 2   Type              7815 non-null   object 
 3   Units             7815 non-null   object 
 4   AFUE              2840 non-null   float64
 5   SEER              4776 non-null   float64
 6   Ducted            4857 non-null   float64
 7   Refrigerant       4845 non-null   float64
 8   Cooling (Btu/hr)  4854 non-null   object 
 9   Heating (Btu/hr)  2919 non-null   object 
 10  MotorType         1170 non-null   object 
 11  HSPF              3459 non-null   object 
 12  FinalEstimate     7815 non-null   object 
 13  EER               3274 non-null   float64
 14  FuelType          2852 non-null   object 
dtypes: float64(5), int64(1), object(9)
memory usage: 915.9+ KB


I renamed all columns for ease of use when using python and modules, and librariaes.

In [11]:
#rename the columns so they are human readable
df=df.rename(columns={'Year':'year','Region':'region','Type':'type','Units':'units','AFUE':'afue','SEER':'seer',
                      'Ducted':'ducted','Refrigerant':'refrigerant','Cooling (Btu/hr)':'cooling_btu_hr',
                      'Heating (Btu/hr)':'heating_btu_hr','MotorType':'motor_type','HSPF':'hspf',
                      'FinalEstimate':'final_estimate','EER':'eer','FuelType':'fuel_type'})



In [12]:
df.head()

year        region type units  afue  seer  ducted  refrigerant  \
0  2021  Northeastern   AC    33   NaN  13.0     0.0        410.0   
1  2021  Northeastern   AC   326   NaN  13.0     0.0        410.0   
2  2021  Northeastern   AC   289   NaN  13.0     0.0        410.0   
3  2021  Northeastern   AC    31   NaN  13.0     0.0        410.0   
4  2021  Northeastern   AC    49   NaN  13.0     0.0        410.0   

  cooling_btu_hr heating_btu_hr motor_type hspf final_estimate  eer fuel_type  
0         18,000            NaN        NaN  NaN            165  NaN       NaN  
1         24,000            NaN        NaN  NaN          1,631  NaN       NaN  
2         30,000            NaN        NaN  NaN          1,446  NaN       NaN  
3         42,000            NaN        NaN  NaN            155  NaN       NaN  
4         48,000            NaN        NaN  NaN            245  NaN       NaN

In [13]:
df.year.value_counts

<bound method IndexOpsMixin.value_counts of 0       2021
1       2021
2       2021
3       2021
4       2021
        ... 
7810    2017
7811    2017
7812    2017
7813    2017
7814    2017
Name: year, Length: 7815, dtype: int64>

In [14]:
df.region.value_counts

<bound method IndexOpsMixin.value_counts of 0       Northeastern
1       Northeastern
2       Northeastern
3       Northeastern
4       Northeastern
            ...     
7810              NY
7811              NY
7812              NY
7813              NY
7814              NY
Name: region, Length: 7815, dtype: object>

In [15]:
df.type.value_counts

<bound method IndexOpsMixin.value_counts of 0            AC
1            AC
2            AC
3            AC
4            AC
         ...   
7810    Furnace
7811    Furnace
7812    Furnace
7813    Furnace
7814    Furnace
Name: type, Length: 7815, dtype: object>

In [16]:
df.units.value_counts

<bound method IndexOpsMixin.value_counts of 0        33
1       326
2       289
3        31
4        49
       ... 
7810      1
7811     15
7812      4
7813      5
7814      8
Name: units, Length: 7815, dtype: object>

In [17]:
df.afue.value_counts

<bound method IndexOpsMixin.value_counts of 0        NaN
1        NaN
2        NaN
3        NaN
4        NaN
        ... 
7810    90.0
7811    90.0
7812    92.0
7813    92.0
7814    92.0
Name: afue, Length: 7815, dtype: float64>

In [18]:
df.seer.value_counts

<bound method IndexOpsMixin.value_counts of 0       13.0
1       13.0
2       13.0
3       13.0
4       13.0
        ... 
7810     NaN
7811     NaN
7812     NaN
7813     NaN
7814     NaN
Name: seer, Length: 7815, dtype: float64>

In [19]:
df.ducted.value_counts

<bound method IndexOpsMixin.value_counts of 0       0.0
1       0.0
2       0.0
3       0.0
4       0.0
       ... 
7810    NaN
7811    NaN
7812    NaN
7813    NaN
7814    NaN
Name: ducted, Length: 7815, dtype: float64>

In [20]:
df.refrigerant.value_counts

<bound method IndexOpsMixin.value_counts of 0       410.0
1       410.0
2       410.0
3       410.0
4       410.0
        ...  
7810      NaN
7811      NaN
7812      NaN
7813      NaN
7814      NaN
Name: refrigerant, Length: 7815, dtype: float64>

In [21]:
df.cooling_btu_hr.value_counts

<bound method IndexOpsMixin.value_counts of 0       18,000
1       24,000
2       30,000
3       42,000
4       48,000
         ...  
7810       NaN
7811       NaN
7812       NaN
7813       NaN
7814       NaN
Name: cooling_btu_hr, Length: 7815, dtype: object>

In [22]:
df.heating_btu_hr.value_counts

<bound method IndexOpsMixin.value_counts of 0           NaN
1           NaN
2           NaN
3           NaN
4           NaN
         ...   
7810    120,000
7811    140,000
7812    100,000
7813    120,000
7814     40,000
Name: heating_btu_hr, Length: 7815, dtype: object>

In [23]:
df.motor_type.value_counts

<bound method IndexOpsMixin.value_counts of 0       NaN
1       NaN
2       NaN
3       NaN
4       NaN
       ... 
7810    ECM
7811    PSC
7812    NaN
7813    NaN
7814    NaN
Name: motor_type, Length: 7815, dtype: object>

In [24]:
df.hspf.value_counts

<bound method IndexOpsMixin.value_counts of 0       NaN
1       NaN
2       NaN
3       NaN
4       NaN
       ... 
7810    NaN
7811    NaN
7812    NaN
7813    NaN
7814    NaN
Name: hspf, Length: 7815, dtype: object>

In [25]:
df.final_estimate.value_counts

<bound method IndexOpsMixin.value_counts of 0               165
1             1,631
2             1,446
3               155
4               245
           ...     
7810     6.67184381
7811    100.0776571
7812    26.68737524
7813    33.35921905
7814    53.37475048
Name: final_estimate, Length: 7815, dtype: object>

In [26]:
df.eer.value_counts

<bound method IndexOpsMixin.value_counts of 0      NaN
1      NaN
2      NaN
3      NaN
4      NaN
        ..
7810   NaN
7811   NaN
7812   NaN
7813   NaN
7814   NaN
Name: eer, Length: 7815, dtype: float64>

In [27]:
df.fuel_type.value_counts

<bound method IndexOpsMixin.value_counts of 0               NaN
1               NaN
2               NaN
3               NaN
4               NaN
           ...     
7810    Natural Gas
7811            Oil
7812    Natural Gas
7813    Natural Gas
7814    Natural Gas
Name: fuel_type, Length: 7815, dtype: object>

In [29]:
df.year.unique()

array([2021, 2018, 2020, 2019, 2017])

In [30]:
df.region.unique()

array(['Northeastern', 'NY'], dtype=object)

In [31]:
df.type.unique()

array(['AC', 'Furnace', 'ASHP', 'Boiler', 'GEU'], dtype=object)

In [32]:
df.units.unique()

array(['33', '326', '289', '31', '49', '28', '2', '3', '16', '20', '19',
       '25', '11', '9', '61', '99', '150', '208', '37', '5', '1', '55',
       '82', '94', '129', '618', '2,112', '1,842', '2,471', '695', '991',
       '590', '263', '469', '572', '694', '231', '211', '120', '118',
       '267', '726', '692', '733', '276', '376', '207', '7', '10', '4',
       '0', '22', '29', '56', '168', '75', '30', '38', '8', '45', '6',
       '17', '87', '114', '843', '12', '362', '187', '15', '136', '117',
       '226', '27', '42', '255', '1,753', '50', '372', '46', '293', '-2',
       '259', '143', '51', '173', '552', '554', '60', '212', '728', '73',
       '152', '206', '128', '63', '1,562', '182', '74', '111', '43',
       '393', '180', '113', '2,369', '84', '575', '112', '85', '58',
       '163', '81', '760', '327', '159', '237', '337', '503', '310', '34',
       '161', '64', '368', '24', '459', '184', '156', '23', '361', '405',
       '216', '53', '21', '35', '14', '13', '-1', '44', '57'

In [33]:
df.afue.unique()

array([nan, 83., 92., 80., 81., 82., 84., 85., 86., 87., 90., 91., 93.,
       94., 95., 96., 97., 98., 88., 89., 99.])

In [34]:
df.seer.unique()

array([13., 14., 15., 16., 17., 18., 20., 22., 23., 19., nan, 21., 24.,
       25., 26.])

In [35]:
df.ducted.unique()

array([ 0.,  1., nan])

In [36]:
df.refrigerant.unique()

array([410.,  nan,  22.])

In [37]:
df.cooling_btu_hr.unique()

array(['18,000', '24,000', '30,000', '42,000', '48,000', '60,000',
       '36,000', '12,000', nan, '6,000', '54,000', '72,000'], dtype=object)

In [38]:
df.heating_btu_hr.unique()

array([nan, '60,000', '40,000', '120,000', '80,000', '100,000', '140,000',
       '160,000', '20,000'], dtype=object)

In [39]:
df.motor_type.unique()

array([nan, 'PSC', 'ECM'], dtype=object)

In [40]:
df.hspf.unique()

array([nan, '7.70-8.19', '8.50-8.99', '8.20-8.49', '10.00-10.49',
       '11.00+', '9.00-9.49', '9.50-9.99', '10.50-10.99'], dtype=object)

In [41]:
df.final_estimate.unique()

array(['165', '1,631', '1,446', ..., '1,687.98', '146.7805638',
       '173.467939'], dtype=object)

In [42]:
df.eer.unique()

array([nan,  9., 12., 20., 10., 11., 13., 14., 15., 16., 18., 19.])

In [43]:
df.fuel_type.unique()

array([nan, 'Oil', 'Natural Gas'], dtype=object)

### I noticed that a few columns need the data type changed to int64, float64, or object:
3   units  int64
7   refrigerant object
8   cooling_btu_hr)int64
9   heating_btu_hr)int64
12  FinalEstimateint64
11  HSPFfloat64

In [ ]:
# update datatypes of binned values to be float
df = df.astype({'sqft_bin': 'float64', 'acres_bin': 'float64', 'age_bin': 'float64',
                    'structure_dollar_sqft_bin': 'float64', 'lot_dollar_sqft_bin': 'float64'})

### I noticed that encoding was needed:
1   Region            7815 non-null   object 
2   Type              7815 non-null   object
7   Refrigerant       4845 non-null   float64
10  MotorType         1170 non-null   object
14  FuelType          2852 non-null   object

# Convert binary categorical variables to numeric
    df['fips_encoded'] = df.fips.map({6037:1, 6059:2, 6111:3})

### I noticed that HSPF needed to be binned due to ranges cells.
11  HSPF              3459 non-null   object





def create_features(df):
    df['age'] = 2017 - df.yearbuilt
    df['age_bin'] = pd.cut(df.age, 
                           bins = [0, 5, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120, 130, 140],
                           labels = [0, .066, .133, .20, .266, .333, .40, .466, .533, 
                                     .60, .666, .733, .8, .866, .933])

In [ ]:
def get_counties():
    '''
    This function will create dummy variables out of the original fips column. 
    And return a dataframe with all of the original columns except regionidcounty.
    We will keep fips column for data validation after making changes. 
    New columns added will be 'LA', 'Orange', and 'Ventura' which are boolean 
    The fips ids are renamed to be the name of the county each represents. 
    '''
    # create dummy vars of fips id
    county_df = pd.get_dummies(df.fips)
    # rename columns by actual county name
    county_df.columns = ['LA', 'Orange', 'Ventura']
    # concatenate the dataframe with the 3 county columns to the original dataframe
    df_dummies = pd.concat([df, county_df], axis = 1)
    # drop regionidcounty and fips columns
    df_dummies = df_dummies.drop(columns = ['regionidcounty'])
    return df_dummies


In [ ]:
# Convert binary categorical variables to numeric
    df['fips_encoded'] = df.fips.map({6037:1, 6059:2, 6111:3})

Summarize the data through aggregates, descriptive stats and distribution plots (histograms, density plots, boxplots, e.g.). (pandas: .value_counts, .head, .shape, .describe, .info, matplotlib.pyplot.hist, seaborn.boxplot)


In [ ]:
def wrangle_hvac():
    '''
    Read hvac into a pandas DataFrame from csv saved on desktop,
    replace whitespaces with NaN values,
    return cleaned student grades DataFrame.
    '''

    # Acquire data

    hvac = get_hvac_data()

    # Replace white space values with NaN values.
    hvac = hvac.replace(r'^\s*$', np.nan, regex=True)

    return df